// @betty-blocks/cli/build/validations/prefab/linkedPartial.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const joi_1 = __importDefault(require("joi"));
const constants_1 = require("../constants");
const linkedPartialValueSchema = joi_1.default.object({
    ref: joi_1.default.object({
        componentId: joi_1.default.string().required(),
    }),
});
exports.linkedPartialConfigurationSchema = joi_1.default.object({
    as: joi_1.default.string().valid(...constants_1.CONFIGURATION_AS),
    condition: joi_1.default.object({
        type: joi_1.default.string().valid(...constants_1.CONDITION_TYPE),
        option: joi_1.default.string(),
        comparator: joi_1.default.string().valid(...constants_1.COMPARATORS),
        value: joi_1.default.any(),
    }),
});
exports.linkedPartialSchema = joi_1.default.object({
    key: joi_1.default.string().required(),
    label: joi_1.default.string().allow(''),
    type: joi_1.default.string().valid('LINKED_PARTIAL').required(),
    value: linkedPartialValueSchema,
    configuration: exports.linkedPartialConfigurationSchema,
});
