// @betty-blocks/cli/build/components/transformers/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const typescript_1 = require("typescript");
const { React } = typescript_1.JsxEmit;
const { ES5 } = typescript_1.ScriptTarget;
exports.transpile = (code, before) => {
    const { diagnostics, outputText } = typescript_1.transpileModule(code, {
        compilerOptions: {
            allowJs: true,
            checkJs: true,
            downlevelIteration: true,
            jsx: React,
            // Requires tslib
            noEmitHelpers: true,
            target: ES5,
        },
        reportDiagnostics: true,
        transformers: {
            before,
        },
    });
    let messageText = '';
    if (diagnostics) {
        diagnostics.forEach((diagnostic) => {
            if (diagnostic.file) {
                const { line, character } = diagnostic.file.getLineAndCharacterOfPosition(diagnostic.start || 0);
                messageText += ` ${diagnostic.file.fileName} (${line + 1},${character + 1})`;
            }
            messageText += `: ${typescript_1.flattenDiagnosticMessageText(diagnostic.messageText, '\n')}`;
        });
    }
    if (messageText) {
        throw new Error(messageText);
    }
    return outputText;
};
exports.doTranspile = (code) => exports.transpile(code, []);
