// @betty-blocks/cli/build/validations/styles.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const chalk_1 = __importDefault(require("chalk"));
const joi_1 = __importDefault(require("joi"));
const types_1 = require("../types");
const shadows = [
    'none',
    '0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12)',
    '0px 3px 1px -2px rgba(0,0,0,0.2),0px 2px 2px 0px rgba(0,0,0,0.14),0px 1px 5px 0px rgba(0,0,0,0.12)',
    '0px 3px 3px -2px rgba(0,0,0,0.2),0px 3px 4px 0px rgba(0,0,0,0.14),0px 1px 8px 0px rgba(0,0,0,0.12)',
    '0px 2px 4px -1px rgba(0,0,0,0.2),0px 4px 5px 0px rgba(0,0,0,0.14),0px 1px 10px 0px rgba(0,0,0,0.12)',
    '0px 3px 5px -1px rgba(0,0,0,0.2),0px 5px 8px 0px rgba(0,0,0,0.14),0px 1px 14px 0px rgba(0,0,0,0.12)',
    '0px 3px 5px -1px rgba(0,0,0,0.2),0px 6px 10px 0px rgba(0,0,0,0.14),0px 1px 18px 0px rgba(0,0,0,0.12)',
    '0px 4px 5px -2px rgba(0,0,0,0.2),0px 7px 10px 1px rgba(0,0,0,0.14),0px 2px 16px 1px rgba(0,0,0,0.12)',
    '0px 5px 5px -3px rgba(0,0,0,0.2),0px 8px 10px 1px rgba(0,0,0,0.14),0px 3px 14px 2px rgba(0,0,0,0.12)',
    '0px 5px 6px -3px rgba(0,0,0,0.2),0px 9px 12px 1px rgba(0,0,0,0.14),0px 3px 16px 2px rgba(0,0,0,0.12)',
    '0px 6px 6px -3px rgba(0,0,0,0.2),0px 10px 14px 1px rgba(0,0,0,0.14),0px 4px 18px 3px rgba(0,0,0,0.12)',
    '0px 6px 7px -4px rgba(0,0,0,0.2),0px 11px 15px 1px rgba(0,0,0,0.14),0px 4px 20px 3px rgba(0,0,0,0.12)',
    '0px 7px 8px -4px rgba(0,0,0,0.2),0px 12px 17px 2px rgba(0,0,0,0.14),0px 5px 22px 4px rgba(0,0,0,0.12)',
    '0px 7px 8px -4px rgba(0,0,0,0.2),0px 13px 19px 2px rgba(0,0,0,0.14),0px 5px 24px 4px rgba(0,0,0,0.12)',
    '0px 7px 9px -4px rgba(0,0,0,0.2),0px 14px 21px 2px rgba(0,0,0,0.14),0px 5px 26px 4px rgba(0,0,0,0.12)',
    '0px 8px 9px -5px rgba(0,0,0,0.2),0px 15px 22px 2px rgba(0,0,0,0.14),0px 6px 28px 5px rgba(0,0,0,0.12)',
    '0px 8px 10px -5px rgba(0,0,0,0.2),0px 16px 24px 2px rgba(0,0,0,0.14),0px 6px 30px 5px rgba(0,0,0,0.12)',
    '0px 8px 11px -5px rgba(0,0,0,0.2),0px 17px 26px 2px rgba(0,0,0,0.14),0px 6px 32px 5px rgba(0,0,0,0.12)',
    '0px 9px 11px -5px rgba(0,0,0,0.2),0px 18px 28px 2px rgba(0,0,0,0.14),0px 7px 34px 6px rgba(0,0,0,0.12)',
    '0px 9px 12px -6px rgba(0,0,0,0.2),0px 19px 29px 2px rgba(0,0,0,0.14),0px 7px 36px 6px rgba(0,0,0,0.12)',
    '0px 10px 13px -6px rgba(0,0,0,0.2),0px 20px 31px 3px rgba(0,0,0,0.14),0px 8px 38px 7px rgba(0,0,0,0.12)',
    '0px 10px 13px -6px rgba(0,0,0,0.2),0px 21px 33px 3px rgba(0,0,0,0.14),0px 8px 40px 7px rgba(0,0,0,0.12)',
    '0px 10px 14px -6px rgba(0,0,0,0.2),0px 22px 35px 3px rgba(0,0,0,0.14),0px 8px 42px 7px rgba(0,0,0,0.12)',
    '0px 11px 14px -7px rgba(0,0,0,0.2),0px 23px 36px 3px rgba(0,0,0,0.14),0px 9px 44px 8px rgba(0,0,0,0.12)',
    '0px 11px 15px -7px rgba(0,0,0,0.2),0px 24px 38px 3px rgba(0,0,0,0.14),0px 9px 46px 8px rgba(0,0,0,0.12)',
];
const isString = joi_1.default.string().max(255);
const validRem = /^\d{1,5}\.?\d{0,5}rem$/;
const optionType = ['STATIC', 'THEME_COLOR'];
exports.contentSchema = joi_1.default.object({
    backgroundColor: joi_1.default.object({
        type: optionType,
        value: isString,
    }),
    borderColor: joi_1.default.object({
        type: optionType,
        value: isString,
    }),
    borderRadius: joi_1.default.array().max(1).items(isString.required().pattern(validRem)),
    borderStyle: isString,
    borderWidth: joi_1.default.array().max(1).items(isString.required().pattern(validRem)),
    boxShadow: shadows,
    color: joi_1.default.object({
        type: optionType,
        value: isString,
    }),
    fontFamily: ['Roboto', 'serif', 'sans-serif', 'monospace'],
    fontSize: isString.pattern(validRem),
    fontStyle: ['italic', 'normal', 'none'],
    fontWeight: ['300', '400', '500', '700'],
    letterSpacing: joi_1.default.alternatives().try(isString.pattern(validRem), isString.allow('normal')),
    lineHeight: joi_1.default.string().max(255).pattern(/^\d*$/),
    padding: [
        isString,
        joi_1.default.array().max(4).items(isString.required().pattern(validRem)),
    ],
    textDecoration: ['underline', 'none'],
    textTransform: ['uppercase', 'none'],
});
const stateSchema = joi_1.default.object({
    name: joi_1.default.string()
        .required()
        .allow(...Object.keys(types_1.AllowedStateKeys)),
    content: exports.contentSchema.required().min(1),
});
const validateStyleType = (componentNames) => (type) => {
    if (componentNames.includes(type)) {
        return type;
    }
    throw new Error(chalk_1.default.red(`\n type for style invalid: ${type} is not available as a component \n`));
};
const styleSchema = (componentNames) => joi_1.default.object({
    name: joi_1.default.string().required(),
    type: joi_1.default.string().required().custom(validateStyleType(componentNames)),
    basis: exports.contentSchema.required().min(1),
    states: joi_1.default.array().items(stateSchema).unique('name').required(),
});
exports.overwriteSchema = joi_1.default.array()
    .items(joi_1.default.object({
    name: joi_1.default.string()
        .required()
        .allow(...Object.keys(types_1.AllowedStateKeys), 'basis'),
    content: exports.contentSchema.required().min(1),
}))
    .unique('name');
const validateUnique = (a, b) => a.name === b.name && a.type === b.type;
const validateAll = (styles, componentNames) => {
    const { error } = joi_1.default.array()
        .items(styleSchema(componentNames))
        .unique(validateUnique)
        .validate(styles);
    if (error) {
        throw new Error(chalk_1.default.red(`\nStyles: ${error.message}\n`));
    }
};
exports.default = (styles, componentNames) => {
    validateAll(styles, componentNames);
};
